Use concordance search in Machinery

abowler2 opened the pull request:
As per the specification for [Concordance search](https://github.com/mozilla/pontoon/blob/master/specs/0106-concordance-search.md), this PR connects to the new concordance search for Machinery searches, uses the project name (or TRANSLATION MEMORY if there is none) in the header, and updates the search field placeholder to "Search Translation Memory".

Of note: If the string appears more than once in memory the project name is repeated (ie `TEST FTL * TEST FTL). I'm not sure if this is expected, but if not then I will need to make changes to account for that.

jotes commented:
>  If the string appears more than once in memory the project name is repeated (ie `TEST FTL * TEST FTL). I'm not sure if this is expected, but if not then I will need to make changes to account for that.

Yeah, The backend [doesn't aggregate duplicated phrases](https://github.com/mozilla/pontoon/pull/1777#discussion_r545764017), but it easily might.

mathjazz reviewed (commented):
Thanks for the patch!

Sadly it doesn't work exactly as specified yet. It replaces "Translation Memory" with "Concordance Search", but instead the default behavior of Machinery should still use Translation Memory. Concordance Search should only be used when search is triggered by manually entering text.

It might be a good idea to consider creating a separate component for Concordance search results given there will be several differences compared to the Machinery results:
1. Project name instead of Translation Memory in the header.
2. No quality or number of hits in the header.
3. [Search match highlighting](https://bugzilla.mozilla.org/show_bug.cgi?id=1684010).
4. No diff between search input and the results.

Regarding grouping results with the same original & translation... We should only print the project name once.

jotes commented:
I'm going to create a patch to aggregate search results and remove
duplicates.

On Sat, 26 Dec 2020 at 00:08, Matjaž Horvat <notifications@github.com>
wrote:

> *@mathjazz* commented on this pull request.
>
> Thanks for the patch!
>
> Sadly it doesn't work exactly as specified yet. It replaces "Translation
> Memory" with "Concordance Search", but instead the default behavior of
> Machinery should still use Translation Memory. Concordance Search should
> only be used when search is triggered by manually entering text.
>
> It might be a good idea to consider creating a separate component for
> Concordance search results given there will be several differences compared
> to the Machinery results:
>
>    1. Project name instead of Translation Memory in the header.
>    2. No quality or number of hits in the header.
>    3. Search match highlighting
>    <https://bugzilla.mozilla.org/show_bug.cgi?id=1684010>.
>    4. No diff between search input and the results.
>
> Regarding grouping results with the same original & translation... We
> should only print the project name once.
> ------------------------------
>
> In frontend/public/static/locale/en-US/translate.ftl
> <https://github.com/mozilla/pontoon/pull/1794#discussion_r548917493>:
>
> > @@ -461,7 +461,7 @@ machinery-Translation--copy =
>  ## Shows the translation source from Pontoon's memory.
>  machinery-TranslationMemory--pontoon-homepage =
>      .title = Pontoon Homepage
> -machinery-TranslationMemory--translation-memory = TRANSLATION MEMORY
> +machinery-TranslationMemory--translation-memory = { $projectName }
>
> We should actually keep this. Only "TRANSLATION MEMORY" should be
> localized (used in default Machinery behavior and in concordance search
> fallback, when project name isn't available), and project names should not
> be localizable.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/mozilla/pontoon/pull/1794#pullrequestreview-558838721>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACDAYTHPACJQGXZE2ZZMAGLSWULPHANCNFSM4VIXA7JQ>
> .
>


-- 
Kind regards

Jarek "jotes" Śmiejczak, a python programmer by day, a mozillian by heart...
Homepage: http://jotes.work, Github: http://github.com/jotes
Mobile: +48693027040, Mozillian: http://mozillians.org/u/jotes

abowler2 commented:
Pushed up the latest changes.  Wanted a bit of input on a few things.
1. Am I on the right track with these changes?  😁 
1. Sorting: ATM, it is still being sorted by the `quality` value. You mentioned that perhaps the Concordance results should be at the top, want to confirm if that is what I should still do.
1. Building off of above, if machinery and concordance have the same quality they both display in the header, but with the styling of concordance (no quality, count, etc..). How should this be addressed? (Perhaps sorting the concordance to the top will 'fix' this as well??)

I think those are the main questions I have ATM. Thanks.

mathjazz commented on frontend/public/static/locale/en-US/translate.ftl:
```diff
@@ -416,7 +416,7 @@ interactivetour-InteractiveTour--end-footer =
 ## Shows a list of translations from machines.
 machinery-Machinery--search-placeholder =
-    .placeholder = Search in Machinery
+    .placeholder = Search Translation Memory
```
This will have to be "Concordance search", sorry spec. We also search other machinery sources, not just TM.

mathjazz reviewed (commented):
1. I'm happy to answer your first question with "Yes". :) You are definitely on the right track. Great progress!

1. I'm afraid we'll actually have to do it vice-versa - other results first, followed by Concordance results. That's because we can't control the number of results for Concordance search, so other results might be hard to reach.

1. (Let's ignore this for now, but we'll also need to do something to prevent JS performance issues - I'm using a prod DB dump locally and can easily get 1000s of results in Concordance search, which slow down the browser a lot while being rendered. Maybe "Load more" button or load on scroll (paging). 

1. The only Concordance search source that has quality assigned is Microsoft Terminology. Which basically always has it set to 100% (because it's actually not quality, but "confidence score" as per the API docs), so we should just remove it from the `getMicrosoftTerminology()`.

abowler2 commented on frontend/src/modules/machinery/components/TranslationMemory.js:
```diff
@@ -25,9 +26,13 @@ export default function TranslationMemory(props: Props) {
+                            <span>TRANSLATION MEMORY</span>
+                        </Localized>
+                    ) : (
+                        <span>{props.projectName.toUpperCase()}</span>
```
I think linking to the translate view would be very nice. Once/if @jotes is able to add those return values, I'll make the change.

mathjazz replied:
It turned out this feature is blocked by https://bugzilla.mozilla.org/show_bug.cgi?id=1324933 (same as https://bugzilla.mozilla.org/show_bug.cgi?id=1390931), because we're grouping multiple TM entries together and we'd need to link to all of them. So we decided to drop the effort within this patch.

Sorry for the confusion!

abowler2 commented:
>     3\. (Let's ignore this for now, but we'll also need to do something to prevent JS performance issues - I'm using a prod DB dump locally and can easily get 1000s of results in Concordance search, which slow down the browser a lot while being rendered. Maybe "Load more" button or load on scroll (paging).

Would you like me to address this now, or in a separate patch?

mathjazz commented on frontend/src/modules/machinery/components/Machinery.js:
```diff
@@ -101,7 +101,7 @@ export default class Machinery extends React.Component<Props> {
                                 type='search'
                                 autoComplete='off'
-                                placeholder='Search in Machinery'
+                                placeholder='Concordance Search'
```
Please also update the FTL file.

mathjazz commented on frontend/src/core/api/machinery.js:
```diff
@@ -160,7 +190,6 @@ export default class MachineryAPI extends APIBase {
                 sources: ['microsoft-terminology'],
                 original: item.source,
                 translation: item.target,
-                quality: Math.round(item.quality),
```
I would also remove `quality` from the backend response.

mathjazz commented on frontend/src/modules/machinery/reducer.js:
```diff
@@ -50,6 +50,9 @@ function dedupedTranslations(
     return translations.sort((a, b) => {
+        if (!a.quality && !b.quality) {
+            return 1;
+        }
```
👍

mathjazz reviewed (commented):
Thanks for the update!

I'd like to avoid merging code that can cause performance issues on prod. So I'd implement paging here. Please sync with @jotes, because we might already be on it.

jotes commented:
Hey, I took a few days off from coding. I'll update my patch on monday to add the pagination.

abowler2 commented:
Pushed up the current WIP status of this branch to get feedback. 

In attempting to `Load More` with the `InfiniteScroll` component I'm getting stuck in an infinite loop. ATM, I've commented out the call to `reset` that occurs in the `machinery.get` function, but if that is left in place then it just resets infinitely and nothing renders. 

I also have some code that I added based on the implementation in `EntitiesList` to see if that was needed to prevent the infinite loop, but it did not work as implemented. I currently have this commented out as well.

Either I didn't implement this correctly, or I need to make the followup call to get additional search results in a different way. Your thoughts on what changes I might need to make would be appreciated.

mathjazz reviewed (commented):
I think I finally made this thing work. I created a patch, which adds back the commented out code, makes a couple of smaller changes highlighted in the comments and then fixes the infinite scroll by making a small CSS change:
https://gist.github.com/mathjazz/2f4bb4f677d071cc9870a39386478fca

The problem was apparently that the parent div of the `InfiniteScroll` component was overflowing the viewport, which was constantly triggering the call to `loadMore`. The CSS change that fixes the issue also make the UI more consistent with the string list - the search box is no longer part of the scrollable panel.

There are at least two more changes that still need to be made:
1. We should not call the `reset()` action when requesting additional pages.
2. We should keep track of the last requested page and only request the next one via the API (if hasMore is true). In the entity list we implement it differently - we make a request to the API with a list of entities already loaded, excluded them, and then always return the first page. Which is weird.

We should also drop this library at some point. :) Probably replace it with something custom made as part of https://bugzilla.mozilla.org/show_bug.cgi?id=1370200#c2.

abowler2 commented:
Pushed up progress, but there is still at least one bug.  The `InfiniteScroll` component does not reset the `pageStart` property unless the component re-renders. As such the scroll only works properly on the first run. If you run another search, clear the search bar, or switch entities the `pageStart` continues to count from where it left off. ~ This is a known issue on their repo with no solution in production at this time.

I made an attempt to track the pages with my own state variable and pass that into `pageStart` but my first run at that caused other problems.

I will continue to work on this, but wanted to push the code to both let you know where I'm at and IF you wanted to take a look yourself you would have access to the current work.

mathjazz commented on frontend/src/modules/machinery/components/source/TranslationMemory.js:
```diff
@@ -25,9 +26,13 @@ export default function TranslationMemory(props: Props) {
+                        </Localized>
+                    ) : (
+                        <span>{props.projectName.toUpperCase()}</span>
+                    )}
```
You can revert this change now that we have `ConcordanceSearch` component.

mathjazz commented on frontend/src/modules/machinery/components/source/ConcordanceSearch.js:
```diff
@@ -0,0 +1,39 @@
+/**
+ * Show the translation source from Pontoon's memory.
+ */
+export default function TranslationMemory(props: Props) {
```
We should change the name of the function.

Maybe also amend the comment a little to differentiate from `TranslationMemory`,

mathjazz reviewed (commented):
Thanks for the update!

I've noticed a couple of minor issues outlined inline.

In addition to that, we need to make sure project name is only printed once for each entry:
<img width="359" alt="Screenshot 2021-01-24 at 18 27 35" src="https://user-images.githubusercontent.com/626716/105638299-31387b80-5e72-11eb-8e00-122da50363a3.png">

I still need to look at the issue with tracking page numbers.

mathjazz commented:
> In addition to that, we need to make sure project name is only printed once for each entry:

As it turns out, the problem is that for each of project names, we always print out the first one:
<img width="359" alt="Screenshot 2021-01-24 at 18 27 35" src="https://user-images.githubusercontent.com/626716/105638299-31387b80-5e72-11eb-8e00-122da50363a3.png">
<img width="411" alt="Screenshot 2021-01-24 at 19 52 18" src="https://user-images.githubusercontent.com/626716/105640345-cab95a80-5e7d-11eb-9d95-89aee395d6d9.png">

abowler2 commented:
> As it turns out, the problem is that for each of project names, we always print out the first one:

Thank you!!  I haven't been able to reproduce with my local data, so that is tremendously helpful 😃

abowler2 commented:
Not sure why the `pytest` is failing. It runs and passes locally.

mathjazz commented on frontend/src/modules/machinery/components/source/TranslationMemory.js:
```diff
@@ -5,6 +5,7 @@ import { Localized } from '@fluent/react';
 
 type Props = {|
     itemCount?: number,
+    projectName?: string,
```
You can remove this one, too.

abowler2 replied:
Gah..... 🤦🏻‍♀️

mathjazz commented on frontend/src/modules/machinery/reducer.js:
```diff
@@ -35,7 +41,8 @@ function dedupedTranslations(
                 newT.original === oldT.original &&
-                newT.translation === oldT.translation,
+                newT.translation === oldT.translation &&
+                newT.projectName === oldT.projectName,
```
IIUC (and testing locally confirms it), we'll show separate entries for each of the items in the example given at https://github.com/mozilla/pontoon/pull/1794#issuecomment-766412395.

abowler2 commented:
OK, so in looking at it again I changed my approach to avoiding the project name being duplicated. However, I'm not convinced this is even the way to go. ATM, it is still rendering the duplicates separately, but prior to this only the first `project_name` was being preserved.  

I'm not sure exactly how to solve this but I have some ideas I will try out. I'll let you know if I get stuck.

abowler2 commented:
All project names are now displaying on one singe entry of a duplicate translation. Let me know if you feel another approach would be better or if there are changes you would like to see in this solution. 

In the meantime, I'll continue to try and figure out how to get the `InfiniteScroll` to reset properly.

mathjazz commented:
Thanks for the update! Grouping works as expected, except in a few places, which I'm unable to reproduce.

So I'm more or less convinced the backend should do the grouping and send us a list of unique source/target combinations with a list of project names. I think it will:
- be easier and prettier to implement in Django than in React,
- perform better and with less jumping,
- consistently render 100 new items with each request,
- reduce network traffic per item,
- and maybe most importantly: allow us to sort items by the number of projects they appear in.

One other thing regarding sorting - we should make sorting by Levenshtein ratio case-insensitive.

@jotes Do you have cycles to take care of these two changes on the backend, i.e.:
1. Return matches as unique source-target pairs with a list of project names they appear in.
2. Use the number of project string appears in as the secondary sorting criteria (after Levenshtein).
3. Ignore case when calculating Levenshtein.ratio().

Regarding paging - I [guess](https://github.com/danbovey/react-infinite-scroller/blob/master/src/InfiniteScroll.js#L44) we'll need to keep track of `page` used in `getMoreResults` in the reducer and reset it when need.

jotes commented:
Hey,
I will take a look tomorrow!

On Mon, 25 Jan 2021 at 22:20, Matjaž Horvat <notifications@github.com>
wrote:

> Thanks for the update! Grouping works as expected, except in a few places,
> which I'm unable to reproduce.
>
> So I'm more or less convinced the backend should do the grouping and send
> us a list of unique source/target combinations with a list of project
> names. I think it will:
>
>    - be easier and prettier to implement in Django than in React,
>    - perform better and with less jumping,
>    - consistently render 100 new items with each request,
>    - reduce network traffic per item,
>    - and maybe most importantly: allow us to sort items by the number of
>    projects they appear in.
>
> One other thing regarding sorting - we should make sorting by Levenshtein
> ratio case-insensitive.
>
> @jotes <https://github.com/jotes> Do you have cycles to take care of
> these two changes on the backend, i.e.:
>
>    1. Return matches as unique source-target pairs with a list of project
>    names they appear in.
>    2. Use the number of project string appears in as the secondary
>    sorting criteria (after Levenshtein).
>    3. Ignore case when calculating Levenshtein.ratio().
>
> Regarding paging - I guess
> <https://github.com/danbovey/react-infinite-scroller/blob/master/src/InfiniteScroll.js#L44>
> we'll need to keep track of page used in getMoreResults in the reducer
> and reset it when need.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mozilla/pontoon/pull/1794#issuecomment-767119921>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACDAYTEEBON4XAV3OQ3N4OLS3XOB7ANCNFSM4VIXA7JQ>
> .
>


-- 
Kind regards

Jarek "jotes" Śmiejczak, a python programmer by day, a mozillian by heart...
Homepage: http://jotes.work, Github: http://github.com/jotes
Mobile: +48693027040, Mozillian: http://mozillians.org/u/jotes

abowler2 commented:
I agree the grouping would probably be handled better on the backend. 

Thank you so much for taking a look at that @jotes

abowler2 commented:
The latest changes make adjustments to account for the change made in the server response. 

In testing that work I was unable to get any additional `sources` to populate my results (I don't have access to the translate APIs and MS Terminology kept failing with error 500). To get around this I 'faked' the return value from MS Terminology and this exposed some behavior changes, most of which I was able to address, that occur when the 'other sources' return results prior to the 'concordance-search'. 

The final issue that I found, and that I'm not quite sure how to handle yet, is that the sort order returned from `concordance-search` is broken when another `source` has a duplicate to that translation. For example, when I search for the word 'File' I find that I have a duplicate translation in 2 of my local projects. When no other `sources` add to the results that entry is displayed at the top of the list with both projects listed in the header. However, if I have that same translation in my mock MS Terminology results then it correctly adds the source name "Microsoft" to the header but the translation is shifted to the bottom of the results when displayed.

When you have a few moments I would love it if you could test this against a larger data set to see if that behavior still occurs under 'normal' use and to see if there are any other issues that I haven't caught.

TIA 😊

mathjazz commented:
Thanks for the update!

Could you please point to the code in which you've addressed "behavior changes" "when the 'other sources' return results prior to the 'concordance-search'"?

To mitigate the problem of mocked results returning earlier than concordance search results, have you tried using [`time.sleep()`](https://docs.python.org/3/library/time.html#time.sleep)?

I've also deployed your branch to stage for real-life testing:
https://mozilla-pontoon-staging.herokuapp.com/sl/firefox/all-resources/?string=74127

It seems like Google Translate always ends up at the bottom, instead of at the top of the results?

abowler2 commented:
> Could you please point to the code in which you've addressed "behavior changes" "when the 'other sources' return results prior to the 'concordance-search'"?
> 

Sorry, that wasn't clear. When one of the `other sources` returns first then the `sources` array will be
```
['<other source>', 'concordance-search']
```
 I've addressed the 'behavior changes' that I could see when that happens (such as when 'concordance-search was no longer at `sources[0]`).  However, I was not able to work out how to handle the sorting issue that I **thought** was related to this.

> To mitigate the problem of mocked results returning earlier than concordance search results, have you tried using [`time.sleep()`](https://docs.python.org/3/library/time.html#time.sleep)?
> 

No, but I will 😁 

> I've also deployed your branch to stage for real-life testing:
> https://mozilla-pontoon-staging.herokuapp.com/sl/firefox/all-resources/?string=74127
> 

Thank you!!! This was very helpful.

Having this on stage also helped me to see that I indeed need to re-add my check to see if the `source` already exists in the array before pushing it in as when paging the `other source` is repeated as many times as there are pages.

> It seems like Google Translate always ends up at the bottom, instead of at the top of the results?

Yes, this is what I was seeing too. Only on my branch it would also display the project names along with the source 'name' but move that result to the bottom as that is where all the 'other sources' end up. (I think.....). What I was/am unsure on is if the ordering of the `sources` array matters or if this will happen regardless.

In addition to the above question: Is this where we want the `other sources` to display on a concordance search?  If so, will they concatenate the way they do on my branch ('Microsoft' - '<project_name>' - '<project_name>'), and if they do is it OK that they then drift to the bottom in that scenario?

mathjazz commented:
We should display other source above the concordance search ([point 2](https://github.com/mozilla/pontoon/pull/1794#pullrequestreview-559745835)). I don't think the order of sources is important, ideally other sources should be (again) listed above projects.

If this turns out to be hard to implement, I think it's fine if we don't "de-duplicate" matches in concordance search.

abowler2 commented:
@mathjazz As you requested, here is the latest version of my work. 

I've added a ref to the InfiniteScroll component in order to be able to call the `resetPage()` function from it. While this solution avoids the errors I was receiving from previous attempts, it does not actually fire the `resetPage()` function as `current` is always `null`

I would appreciate any thoughts you have on how to get access to the reset function in order to utilize it when clearing the search input or changing entities (currently the page value does not get reset within the InfiniteScroll unless the component re-renders).

abowler2 commented:
OK, concordance search with a `Load more` button for your review 😅

julen commented on frontend/src/modules/machinery/components/Machinery.js:
```diff
@@ -47,18 +52,27 @@ export default class Machinery extends React.Component<Props> {
     };
 
+    getMoreResults = () => {
+        const nextPage = this.state.page + 1;
```
When setting state based on a current state value, please note you should do that in an updater function, [because state updates may be asynchronous](https://reactjs.org/docs/state-and-lifecycle.html#state-updates-may-be-asynchronous). So your code should read:
```js
this.setState((state) => ({
    page: state.page + 1,
}));
```

_(Please also note this may look a bit different on the surface when using hooks, e.g. `const [page, setPage] = useState(0); ... setPage(page + 1);`)_

You may ask, what about `nextPage` two lines below, then? That code is meant to go in a callback function as the second parameter to `setState()`, where you have access to the updated state (hence `this.state.page` will point to the _next_ page), although it could as well be written as part of the `componentDidUpdate` lifecycle method. [The docs recommend the latter](https://reactjs.org/docs/react-component.html#setstate), actually:
> The second parameter to `setState()` is an optional callback function that will be executed once `setState` is completed and the component is re-rendered. Generally we recommend using `componentDidUpdate()` for such logic instead.

Either way it'll be fine. The important thing is to understand the potential for asynchronous updates and how to deal with them.

abowler2 replied:
Setting the value increment on `const nextPage` and not using the state directly was my approach to handling the asynchronicity of state updates. Albeit, I wasn't convinced it was the best approach 😁 

Thanks for the resource links as they were really helpful in connecting the reasoning behind your thoughts. I'll go back over this and see if I can clean it up a bit.

abowler2 replied:
I've been unable to modify this code without raising additional issues.

If I just increment the page number in the `getMoreResults` function using the updater function for `setState` and then call `searchMachinery` in `componentDidUpdate` based on if `this.state.page !== prevState.page` I get the correct behavior for loading the additional search results. However, the functions in the `actions` file look for the existence of `page` to determine if they are to reset the overall state and this is not properly updated within this flow so the `Load More` button remains in certain instances.

If I keep the call to `searchMachinery` in the `getMoreResults` function then state has not had a chance to update and I do not get the proper value for page within the call.

Currently, I am taking the current value of `this.state.page`, adding 1, and storing that value in a variable that I then use to both update the state of `page` and call to `searchMachinery`.  This flow is successful in all scenarios that I need for this component and as such I will leave it for now.

Feel free to let me know if I am just missing something as I'm happy to revisit this if that is the case.

julen replied:
If I understood correctly the problem you were facing, resetting the value of `machinery.hasMore` in the reducer as part of handling the `RESET` action seems to be the natural thing to do here. I just tried this locally with the latest state of this PR and it works fine, although I may be missing some details or relevant context.

abowler2 replied:
Yes, I had overlooked that previously and I've made that adjustment on my local branch now. 

However, the problem I encountered when calling to `searchMachinery` from `componentDidUpdate` was that after the `Load more` button has been clicked, if I attempt to reset the search by clicking the  `x` button, the call to `searchMachinery` is triggered with the reset of `page` to the starting value of 1 and the first page of results is returned along with t`hasMore` as true and so the button again shows.

Not sure if that made sense, so here is a screencast: 

https://user-images.githubusercontent.com/19564687/107890727-4469c480-6ecf-11eb-8ea1-14e5c138e7b2.mov

julen replied:
Thanks for the screencast, I see much better now what problem you are running into!

> then call searchMachinery in componentDidUpdate based on if `this.state.page !== prevState.page`

This should only apply if it's not the initial page, i.e. `this.state.page !== 1`, because we don't want a `machinery.reset` action to trigger this behavior.

For the scenario you are dealing with here, I believe it's probably easier to reason about the code when using the callback to `setState`:

```js
const inputValue = this.searchInput.current.value;
this.setState(
    (state) => ({ page: state.page + 1 }),
    () => {
        this.props.searchMachinery(inputValue, this.state.page);
    }
);
```

I couldn't hit any of the bugs you mentioned with this code above (nor with the `this.state.page !== 1` check in `componentDidUpdate`).

julen commented on frontend/src/modules/machinery/components/source/ConcordanceSearch.js:
```diff
@@ -0,0 +1,39 @@
+                id='machinery-ConcordanceSearch--pontoon-homepage'
+                attrs={{ title: true }}
+            >
+                <a
```
How is an anchor element useful here if its action is prevented?

mathjazz replied:
We just prevent propagation in order to prevent copying the content into the editor when clicking on header.

But I do think we should drop these links - no point in linking to Pontoon Homepage - especially in the light of this proposal: https://github.com/mozilla/pontoon/pull/1794#issuecomment-778156337.

julen commented on frontend/src/modules/machinery/reducer.js:
```diff
@@ -72,13 +92,23 @@ const initial: MachineryState = {
+        case CONCORDANCE_SEARCH:
+            return {
+                ...state,
+                searchResults: state.searchResults.concat(action.searchResults),
```
Use whatever you find more intuitive, but given that object spread is already used one line above, you may write this using the array spread operator, too:
```js
searchResults: [...state.searchResults, ...action.searchResults],
```
Both `concat()` and array spread return a new array and are interchangeable.

abowler2 replied:
As with the comment below, this is another instance of following the existing pattern within the code base, and using the same format as that used for the Entities list.

mathjazz replied:
OTOH, since there's just a handful of cases like this in the codebase (or even just one?), you could change it as part of this PR if you prefer the spread operator. :trollface:

julen commented on frontend/src/modules/machinery/components/Translation.js:
```diff
@@ -79,39 +81,52 @@ export default function Translation(props: Props) {
+                ) : (
+                    <>
+                        <header>
+                            {!translation.quality ? null : (
```
Can `translation.quality` be `0`? If not, then `translation.quality && <span...>` should suffice, i.e. no need to check for the negative condition.

mathjazz replied:
We use the `!condition ? null : component` pattern a lot in our code. I don't have strong opinions about the code style, and I also find the proposed one cleaner. I appreciate consistency, though.

So if you go ahead with the proposed change, please also file a bug to replicate the change on other instances of the pattern.

abowler2 replied:
Yes, the formatting here was in keeping with the existing pattern within the code. However, I also prefer the `&&` format and use it extensively in my other work (in fact, I realized that I used it when displaying the button in this MR without thinking 😁 ).

I will go forward with the suggested format and file the bug to go back and replace the existing instances as I agree that consistency is important.

julen commented on frontend/src/modules/machinery/components/Machinery.js:
```diff
@@ -90,24 +106,48 @@ export default class Machinery extends React.Component<Props> {
+                        {machinery.searchResults.map((result, index) => {
+                            return (
+                                <Translation
+                                    index={index}
```
I haven't tested this myself but I think this will create a problem with the `ctrl+shift+up/down` keyboard shortcuts, which rely on indexes to select a prev/next result from helpers. The idea is that indexes should be consecutive, and they shouldn't overlap. To fix this, you should be able to offset this index with the amount of machinery translations.

Additionally, the value of `key` can also create other issues ([see how React uses the `key` prop](https://reactjs.org/docs/lists-and-keys.html#keys)), especially because (and correct me if I'm wrong) search results are sorted as they arrive. Do we have a uniquely-identifying value for every search result that we could use as a `key`?

abowler2 replied:
I didn't love using the index for the key, but sadly the only possible 'unique' value will be the translation which can be very long in some cases, and I'm also not sure if it will always be unique as the items are only combined into a single value if both the original string and the translation are the same. I suppose there could be times when the original might be different but the translation would be the same.

With that in mind I will offset the index value to avoid conflict with the shortcut.

abowler2 replied:
After taking a closer look here, the use of `index` was how this was originally set up when only mapping the `translations`. So, I tested the shortcut you mentioned, and it still works as expected.

julen replied:
I tested this on my end and I'm afraid it doesn't work as expected 😬. Please check the following screencast, where all taken shortcut actions are `ctrl+shift+down`.

https://user-images.githubusercontent.com/16768/107953295-3d0fed00-6f9b-11eb-8975-53bc0cc02418.mov

You'll notice 2 bugs:
* Performing concordance search leaves the element at previously-selected index highlighted. The easy fix would be to reset the selected index upon search (`editor.actions.resetHelperElementIndex()`).
* Cycling through search results doesn't respect visual ordering. I'm almost certainly sure this is related to the original `index` problem I was mentioning earlier here.

abowler2 replied:
Ah, yes...... I fully see the problem now.  Thanks for your patience 😊

mathjazz commented:
Nice work April!

Just a few high-level observations for now:
* The button remains visible even in the default Machinery view, after you switch entity after performing concordance search.
* The button disappears after you click on it, causing the UI to jump. Ideally we should add a loader after the request returns, same as we do with the Save/Suggest button.
* The list of projects gets really long sometimes. While the spec doesn't cover it, I think we should only put the list of projects in 1 line, end it with an ellipsis and show a full list of projects on hover (anywhere in the header). I can also check how expensive would it be to put projects in a more meaningful order on backend (e.g. by enabled status and priority).
<img width="360" alt="Screenshot 2021-02-12 at 13 00 33" src="https://user-images.githubusercontent.com/626716/107765555-5cf0a800-6d32-11eb-8da3-46cc9b83164b.png">

abowler2 commented on frontend/src/modules/machinery/components/Machinery.js:
```diff
@@ -127,32 +125,28 @@ export default class Machinery extends React.Component<Props> {
+                                    className='load-more-button'
+                                    onClick={this.getMoreResults}
+                                >
+                                    Load More
```
😭  I intended to go back and localize this but clearly forgot to before pushing.

I'll address this along with the other items you mentioned above.

abowler2 commented:
> The list of projects gets really long sometimes. While the spec doesn't cover it, I think we should only put the list of projects in 1 line, end it with an ellipsis and show a full list of projects on hover (anywhere in the header). I can also check how expensive would it be to put projects in a more meaningful order on backend (e.g. by enabled status and priority).

@mathjazz After taking a look at this, I have a question.

The project list that renders is generated as the result of a mapping function. Currently the title is added to each individual project name as it is mapped to an `li`. Once the map is completed it is rendered with the appropriate styling that adds the separator dot between the `li`s. 

---
Should I try to work out a way to have the whole list display when hovering over any of the displayed project names?  Currently, I don't have access to all the projects when creating the title, so I would have to pass that info in and then style it accordingly. 

~ OR ~ 

Should I add the title to the header element?  Doing this interferes with the current hover action that is set to the entire `div` and displays the title for copying into the translation, meaning that the copy title won't display if they hover over the header section but rather the list of projects will.

mathjazz commented:
> Should I try to work out a way to have the whole list display when hovering over any of the displayed project names? Currently, I don't have access to all the projects when creating the title, so I would have to pass that info in and then style it accordingly.

I'm afraid this option will result in jumping on hover...

> Should I add the title to the header element? Doing this interferes with the current hover action that is set to the entire `div` and displays the title for copying into the translation, meaning that the copy title won't display if they hover over the header section but rather the list of projects will.

So if my assumption is correct, I'd go with this option. For most of the header (i.e. Machinery sources) we already show different titles.

abowler2 commented:
I believe I have addressed all feedback, but please be sure to let me know if I missed something or if more changes are needed.

Also, I didn't rebase to resolve the conflicts yet as @mathjazz's pending [PR for dropping caighdean](https://github.com/mozilla/pontoon/pull/1871) conflicts as well . Given that it was approved I will wait to rebase until that has been merged.

abowler2 commented:
>For most of the header (i.e. Machinery sources) we already show different titles.

Yeah, when I first looked into how to solve this I was thinking I had to add it to the actual `header` element which would behave differently than all the other titles, but I was able to work out how to add it to the `ul` and as such it behaves like the other sources in regards to displaying the title.

mathjazz commented:
> Also, I didn't rebase to resolve the conflicts yet as @mathjazz's pending [PR for dropping caighdean](https://github.com/mozilla/pontoon/pull/1871) conflicts as well . Given that it was approved I will wait to rebase until that has been merged.

Merged. :)

abowler2 commented:
_Flow's lack of support for optional chaining......_

![](https://media3.giphy.com/media/1lAOemoi0KhPMzxczT/giphy.gif)



Merge conflicts addressed and all tests passing.

mathjazz commented:
I only found one issue!

<img width="213" alt="Screenshot 2021-02-16 at 23 19 17" src="https://user-images.githubusercontent.com/626716/108128689-6e78dd80-70ad-11eb-868d-091ea6617058.png">

Which gave me an idea: we should only print "Translation Memory" if we don't have any projects in the list.

mathjazz commented on frontend/src/modules/machinery/components/TranslationSource.js:
```diff
@@ -51,5 +68,31 @@ export default function TranslationSource({ translation, locale }: Props) {
+        }
+        if (translation.projectNames.every((projectName) => !projectName)) {
+            return 'Translation Memory';
+        }
```
You'd need to make that localizable... but we'll only show "Translation Memory" alone (if there aren't any projects), so I'd instead either remove the tooltip or probably even better - reuse the `TranslationMemory` component.

mathjazz reviewed (commented):
This works really well now! Thumbs up!

Left some comments for potential code improvements.

abowler2 commented on frontend/src/core/api/machinery.js:
```diff
@@ -20,6 +24,42 @@ export default class MachineryAPI extends APIBase {
+        const params = {
+            text: source,
+            locale: locale.code,
+            page: page ? page.toString() : 1,
```
I'm not able to get logical OR to work here as I can't call `toString()` on page without first confirming if page exists (if Flow allowed `optional chaining` then I could use the logical OR  here). 

Would you still like me to cast the `1` to a string? If so, do you have a preference for doing that (string literal, toString method, template literal...?

mathjazz commented on frontend/src/modules/entitydetails/components/EntityDetails.js:
```diff
@@ -227,9 +227,21 @@ export class EntityDetailsBase extends React.Component<InternalProps, State> {
+            );
+        }
+
+        if (pk || !page) {
```
I'm curious, why did you also add `pk ||`? Doesn't seem necessary.

To add more clarity, perhaps we can split this method into `searchMachinery` (concordance search, including loading more results) and `resetMachinery` (reset machinery with a click or when searching empty string). I'll leave the decision to you.

abowler2 replied:
TBH, I don't know why I thought that was needed :(

I think splitting it into 2 methods would add clarity. For now I've removed `pk` from the condition, and I will explore splitting this out a bit later today.

mathjazz commented on frontend/src/modules/machinery/components/ConcordanceSearch.js:
```diff
@@ -0,0 +1,83 @@
+
+    const getProjectNames = () => {
+        if (!translation.projectNames) {
+            return undefined;
```
Can this be `null` instead?

mathjazz reviewed (commented):
Nice work!

I see this error when performing concordance search in the JS Console:

```
Warning: Each child in a list should have a unique "key" prop.

Check the render method of `ConcordanceSearch`. See https://fb.me/react-warning-keys for more information.
    in li (at ConcordanceSearch.js:34)
    in ConcordanceSearch (at Translation.js:85)
    in li (at Translation.js:78)
    in Localized (at Translation.js:77)
    in Translation (at Machinery.js:136)
    in ul (at Machinery.js:133)
    in div (at Machinery.js:120)
    in section (at Machinery.js:93)
    in Machinery (at Helpers.js:150)
    in div (created by TabPanel)
    in TabPanel (at Helpers.js:149)
    in div (created by UncontrolledTabs)
    in UncontrolledTabs (created by Tabs)
    in Tabs (at Helpers.js:126)
    in div (at Helpers.js:125)
    in Helpers (at EntityDetails.js:483)
    in section (at EntityDetails.js:482)
    in section (at EntityDetails.js:441)
    in EntityDetailsBase (at EntityDetails.js:548)
    in EntityDetails (at App.js:128)
    in section (at App.js:126)
    in div (at App.js:106)
    in App (created by ConnectFunction)
    in ConnectFunction (at src/index.js:30)
    in LocalizationProvider (at AppLocalizationProvider.js:49)
    in AppLocalizationProviderBase (created by ConnectFunction)
    in ConnectFunction (at src/index.js:29)
    in Router (created by ConnectedRouter)
    in ConnectedRouter (created by Context.Consumer)
    in ConnectedRouterWithContext (created by ConnectFunction)
    in ConnectFunction (at src/index.js:28)
    in Provider (at src/index.js:27)
```

mathjazz reviewed (approved):
This looks great, let's ship it! :)

Thank you for your hard work April, and sorry for all the deviations from the spec.

As I already mentioned, I'd like to order projects by priority (and put disabled to the end of the list). I'll make a separate PR for that.

mathjazz commented:
Forgot the GIF! I'm slacking...

mathjazz commented:
![](https://media.giphy.com/media/8FDdXCyA5MSJs3qT5k/giphy.gif)
